draco/analysis/transform.py

```diff
@@ -1265,3 +1266,299 @@ def process_finish(self) -> Union[containers.SiderealStream, containers.RingMap]
+            # Apply downselections and redistribute for reduction
+            weight = _downselect(data.weight)
+
+            if self.weight_op in {"all", "any"}:
+                # Convert to boolean before applying the op
+                w = weight > 0.0
+            else:
+                w = weight
+
+            # Apply the weight reduction
+            ds_weight = self._weight_op_func(w, axis=apply_over, keepdims=True)
+
+            if out.weight.distributed:
+                # Distribute back to the original axis
+                ds_weight = ds_weight.redistribute(original_ax_id)
+
+            out.weight[:] = ds_weight[:]
+        else:
+            # Flat weight that can broadcast to whatever shape
+            self.log.info("No weights available. Using equal weighting.")
+            weight = np.ones(1, dtype=bool)
+
+        # Iterate over the datasets and reduce
+        for name in self.datasets:
+            # Get the dataset
+            ds = data.datasets[name]
+            # Initialize the dataset in the output container
+            out.add_dataset(name)
+            # Get the current distributed axis of this dataset, if it
+            # is distributed
+            if ds.distributed:
+                original_ax_id = ds.distributed_axis
+
+            # Get the axes for this dataset
+            ds_axes = ds.attrs["axis"]
+            # Get the axes in this dataset to reduce over
+            apply_over = _select_red_axes(ds_axes)
+            # Apply downselections and redistribute
+            arr = _downselect(ds)
+            # If arr is distributed, get its axis and a numpy view
+            if isinstance(arr, mpiarray.MPIArray):
+                # Get the new distributed axis
+                new_ax = arr.axis
+                arr = arr.local_array[:]
+
+            if name in self.apply_weight or self.mask_array:
+                # If we need to use the weights, make sure they're distributed
+                # to the correct axis and can broadcast to this array
+                if isinstance(weight, mpiarray.MPIArray):
+                    # Redistribute to the equivalent axis in the weights
+                    ax_name = ds_axes[new_ax]
+                    new_weight_ax = weight_axes.index(ax_name)
+                    weight = weight.redistribute(new_weight_ax)
+
+                # Broadcast the weights to the array shape
+                # MPIArrays aren't preserved properly, so must return a ndarray
+                try:
+                    mask = np.broadcast_to(weight, arr.shape, subok=False)
+                except ValueError as e:
+                    raise ValueError(
+                        f"Could not broadcast weights. Got dataset {name} with "
+                        f"shape {arr.shape} and weight with shape {weight.shape}."
+                    ) from e
+
+            if name in self.apply_weight:
+                # Apply the weights to the dataset
+                arr = arr * mask
```

Comment:
I think this weighting isn't quite what we want. For reductions that are essentially summation you want to do something like:
```
arr = arr * weight / weight.sum(axis=apply_over) * arr.shape[apply_over]
```

Reply:
I've cleaned this up a fair bit. I'm not 100% sure on the weighting for a multi-axis variance, so it would helpful to check that.